### ISSUE TITLE
[Feature] MLA decode: add fold_sq to trtllm-gen so one backend is fast at all context lengths

### ISSUE THREAD

**Issue by yhyang201** [flashinfer-ai/flashinfer#4390]
## TL;DR

All numbers below are attention TP8, DCP off.

On SM100, MLA decode has two backends: `trtllm-gen` and `cute-dsl`. Neither is faster in all cases. `trtllm-gen` wins at `q=1` (plain decode) and at short context. `cute-dsl` wins at `q>1` (MTP/spec verify) above a context threshold (~100k for bf16 q=8, ~200k for fp8). A framework must fix one backend at launch (it goes into the CUDA graphs), and `backend="auto"` + the autotuner cannot pick by context length. We ask for `fold_sq` in `trtllm-gen` so a single backend is fast at every context length.

## What we see

On SM100 (B200/B300), neither MLA decode backend is faster in all cases:

- `q=1` (plain decode): `trtllm-gen` always faster.
- `q>1` (MTP/spec verify): `cute-dsl` faster above a context threshold (~100k for bf16 q=8, ~200k for fp8), `trtllm-gen` below it.

Example, TP8, µs/call (trtllm-gen | cute-dsl), decode backend the only variable (lower is faster):

| q | 131k | 500k | 1000k |
|---|---|---|---|
| 1 (bf16) | 94 / 122 | 143 / 177 | 254 / 331 |
| 8 (bf16) | 201 / 124 | 717 / 193 | 1413 / 359 |
| 8 (fp8) | 103 / 124 | 343 / 123 | 667 / 168 |

Reason: `cute-dsl` folds the `q` verify tokens into the MMA tile (`fold_sq`), so it reads the KV once and stays flat in context. `trtllm-gen` re-reads KV per query row, so it grows as O(ctx·q). End-to-end in SGLang (Kimi-K3, q=8, 900k, 8×B300): decode throughput 98 → 197 tok/s.

## Why the autotuner doesn't help

`backend="auto"` builds both runners, but it cannot pick by context length:
- It tunes at one fixed context (`profile_seq_len`) and sweeps the batch dim, not context ([`mla/_core.py` L2152](https://github.com/flashinfer-ai/flashinfer/blob/v0.6.15.post1/flashinfer/mla/_core.py#L2152), [L2198](https://github.com/flashinfer-ai/flashinfer/blob/v0.6.15.post1/flashinfer/mla/_core.py#L2198)).
- It profiles only inside `flashinfer.autotune(True)`; otherwise it returns `trtllm-gen` with no measurement.

A launch-time choice does not work either. The best backend depends on the run-time context length. An operator does not know that length at launch. One deployment serves many context lengths, and the context of a single sequence grows during decode. So no fixed choice is right for every request. The operator cannot pick the best backend, and SGLang cannot pick it for them, because the backend goes into the CUDA graphs and cannot change per request or per step.

## Why it hurts (MTP)

Without DCP, MTP mixes `q=1` and long-context `q>1` steps and wants one backend for both. No single backend wins: `trtllm-gen` is faster at `q=1`, `cute-dsl` is faster at long-context `q>1`. The backend is fixed at launch (it goes into the CUDA graphs), so the framework must eat a regression on one regime.

## Request

Add `fold_sq` to the `trtllm-gen` MLA decode kernel (fold `seq_q` into MMA-M, as `cute-dsl` does). It stays best at `q=1` and becomes flat in context at `q>1`, so `trtllm-gen` wins or ties at every context length. A framework can then standardize on one backend with no context-length trade-off.

## Reproduction

flashinfer `0.6.15.post1`, one B300 (SM100). The script sweeps the isolated MLA decode call for both backends and prints the table above (TP8 / H=12, q=1 and q=8, bf16 and fp8, ctx 131k/500k/1000k), µs per call.

<details><summary>repro.py</summary>

```python
import torch, flashinfer

dev = "cuda"
KV_LORA, ROPE, NOPE, HDQK, PAGE = 512, 64, 128, 576, 64
H = 12  # query heads per rank at TP8 (1 KV head, MQA)
torch.manual_seed(0)

def counter_buf(h):
    import flashinfer.utils as fu
    try:
        n = int(fu.get_trtllm_gen_multi_ctas_kv_counter_bytes(dev, h, 1))
        return torch.zeros(n, dtype=torch.uint8, device=dev)
    except Exception:
        return None

def make_kv(kv_len, dt):
    npp = (kv_len + PAGE - 1) // PAGE
    if npp % 2:
        npp += 1
    kv = torch.randn(npp, PAGE, HDQK, device=dev, dtype=torch.bfloat16).to(dt)
    bt = torch.arange(npp, device=dev, dtype=torch.int32).view(1, npp)
    sl = torch.full((1,), kv_len, device=dev, dtype=torch.int32)
    ws = torch.zeros(512 * 1024 * 1024, device=dev, dtype=torch.uint8)
    return kv, bt, sl, ws

def bench(kv, bt, sl, kv_len, q, backend, ws, dt):
    query = torch.randn(1, q, H, HDQK, device=dev, dtype=torch.bfloat16).to(dt)
    kw = dict(query=query, kv_cache=kv, workspace_buffer=ws,
              qk_nope_head_dim=NOPE, kv_lora_rank=KV_LORA, qk_rope_head_dim=ROPE,
              block_tables=bt, seq_lens=sl, max_seq_len=kv_len,
              bmm1_scale=1 / (192 ** 0.5), bmm2_scale=1.0)
    if backend == "trtllm-gen":
        kw.update(backend="trtllm-gen", enable_pdl=True,
                  multi_ctas_kv_counter_buffer=counter_buf(H))
    else:
        kw.update(backend=backend)
    f = flashinfer.decode.trtllm_batch_decode_with_kv_cache_mla
    try:
        for _ in range(8):
            f(**kw)
    except Exception:
        return None
    torch.cuda.synchronize()
    s, e = torch.cuda.Event(True), torch.cuda.Event(True)
    s.record()
    for _ in range(30):
        f(**kw)
    e.record()
    torch.cuda.synchronize()
    return s.elapsed_time(e) / 30 * 1000  # us/call

print("dtype,q,ctx,trtllm_us,cutedsl_us")
for dt, tag in [(torch.bfloat16, "bf16"), (torch.float8_e4m3fn, "fp8")]:
    for q in [1, 8]:
        for kv_len in [131072, 500000, 1000000]:
            kv, bt, sl, ws = make_kv(kv_len, dt)
            tr = bench(kv, bt, sl, kv_len, q, "trtllm-gen", ws, dt)
            cd = bench(kv, bt, sl, kv_len, q, "cute-dsl", ws, dt)
            print(f"{tag},{q},{kv_len},{tr and round(tr,1)},{cd and round(cd,1)}", flush=True)
            del kv, bt, sl, ws
            torch.cuda.empty_cache()
```

</details>

Happy to share the wider sweep (TP1/TP4/TP16, all q, more contexts) and nsys captures.

**Comment by Observer007** [comment]
cc @PerkzZheng

**Comment by PerkzZheng** [comment]
@yhyang201 about “and backend="auto" + the autotuner cannot pick by context length”, can you explain a bit more details about this ？

Also, we are open-sourcing [another attention backend](https://github.com/flashinfer-ai/flashinfer/pull/4357) (which should have relatively good performance in both sq = 1 and sq > 1) that supports groups tokens and heads ? can you share the benchmarking scripts you are using and we can give it a quick try with those new kernels.

**Comment by nvpohanh** [comment]
> A framework must fix one backend at launch (it goes into the CUDA graphs), and backend="auto" + the autotuner cannot pick by context length. 

In Decode, we capture one cuda graph per num_tokens. So we must decide on one backend for each num_tokens.

But the current situation is that at a specific num_tokens, the best tactic choice depends on the context length (i.e. history kv length), so regardless of which tactic we choose, it is always a suboptimal choice for some context lengths at this num_tokens.

The fundamental issue is that we can only capture cuda graph following one dimension (i.e. num_tokens) but not two dimensions (i.e. num_tokens x context_length) because in that case the number of captured cuda graphs will explode :(

@yhyang201 let me know if I explained it incorrectly. Thanks!

**Comment by Observer007** [comment]
Hi, @yhyang201  Is the performance data number you listed measured by cuda graph? It is a bit weird that in fp8 500k performs better than 131k, below data is measured on B200. 
| KV dtype | Q length | Context length | TRTLLM-Gen Eager | TRTLLM-Gen CUDA Graph | CuTe DSL Eager | CuTe DSL CUDA Graph |
  |---|---:|---:|---:|---:|---:|---:|
  | BF16 | 1 | 131,072 | 62.4 | 50.8 | 101.7 | 55.3 |
  | BF16 | 1 | 500,000 | 135.8 | 139.7 | 185.3 | 185.9 |
  | BF16 | 1 | 1,000,000 | 264.3 | 265.2 | 364.2 | 364.4 |
  | BF16 | 8 | 131,072 | 206.4 | 205.1 | 100.7 | 62.4 |
  | BF16 | 8 | 500,000 | 725.6 | 728.2 | 201.5 | 204.4 |
  | BF16 | 8 | 1,000,000 | 1,439.7 | 1,440.4 | 394.6 | 393.3 |
  | FP8 | 1 | 131,072 | 63.0 | 31.8 | 101.8 | 32.8 |
  | FP8 | 1 | 500,000 | 70.6 | 72.9 | 102.0 | 95.9 |
  | FP8 | 1 | 1,000,000 | 129.2 | 131.6 | 197.7 | 200.4 |
  | FP8 | 8 | 131,072 | 105.6 | 108.6 | 101.4 | 36.6 |
  | FP8 | 8 | 500,000 | 356.3 | 360.9 | 105.5 | 108.3 |
  | FP8 | 8 | 1,000,000 | 745.5 | 752.4 | 204.1 | 206.0 |

cc @PerkzZheng @nvpohanh

**Comment by PerkzZheng** [comment]
> > A framework must fix one backend at launch (it goes into the CUDA graphs), and backend="auto" + the autotuner cannot pick by context length.
> 
> In Decode, we capture one cuda graph per num_tokens. So we must decide on one backend for each num_tokens.
> 
> But the current situation is that at a specific num_tokens, the best tactic choice depends on the context length (i.e. history kv length), so regardless of which tactic we choose, it is always a suboptimal choice for some context lengths at this num_tokens.
> 
> The fundamental issue is that we can only capture cuda graph following one dimension (i.e. num_tokens) but not two dimensions (i.e. num_tokens x context_length) because in that case the number of captured cuda graphs will explode :(
> 
> [@yhyang201](https://github.com/yhyang201) let me know if I explained it incorrectly. Thanks!

we would not expect that the best kernel selection changes as seqlen_kv increases (at least the gaps should not be such large). As @Observer007 measured, it makes sense to select attention kernels only based on seqlen_q then.

**Comment by yhyang201** [comment]
> Hi, [@yhyang201](https://github.com/yhyang201) Is the performance data number you listed measured by cuda graph? It is a bit weird that in fp8 500k performs better than 131k, below data is measured on B200.
> 
> KV dtype	Q length	Context length	TRTLLM-Gen Eager	TRTLLM-Gen CUDA Graph	CuTe DSL Eager	CuTe DSL CUDA Graph
> BF16	1	131,072	62.4	50.8	101.7	55.3
> BF16	1	500,000	135.8	139.7	185.3	185.9
> BF16	1	1,000,000	264.3	265.2	364.2	364.4
> BF16	8	131,072	206.4	205.1	100.7	62.4
> BF16	8	500,000	725.6	728.2	201.5	204.4
> BF16	8	1,000,000	1,439.7	1,440.4	394.6	393.3
> FP8	1	131,072	63.0	31.8	101.8	32.8
> FP8	1	500,000	70.6	72.9	102.0	95.9
> FP8	1	1,000,000	129.2	131.6	197.7	200.4
> FP8	8	131,072	105.6	108.6	101.4	36.6
> FP8	8	500,000	356.3	360.9	105.5	108.3
> FP8	8	1,000,000	745.5	752.4	204.1	206.0
> cc [@PerkzZheng](https://github.com/PerkzZheng) [@nvpohanh](https://github.com/nvpohanh)

Hi @Observer00. These were eager runs, not CUDA graph. At ~900k the eager numbers matched what we see end-to-end in SGLang with graphs, so I kept the sweep simple and skipped graph capture.  Didn't expect eager vs graph to diverge this much at 131k, thanks for flagging it.

**Comment by Observer007** [comment]
> > > A framework must fix one backend at launch (it goes into the CUDA graphs), and backend="auto" + the autotuner cannot pick by context length.
> > 
> > 
> > In Decode, we capture one cuda graph per num_tokens. So we must decide on one backend for each num_tokens.
> > But the current situation is that at a specific num_tokens, the best tactic choice depends on the context length (i.e. history kv length), so regardless of which tactic we choose, it is always a suboptimal choice for some context lengths at this num_tokens.
> > The fundamental issue is that we can only capture cuda graph following one dimension (i.e. num_tokens) but not two dimensions (i.e. num_tokens x context_length) because in that case the number of captured cuda graphs will explode :(
> > [@yhyang201](https://github.com/yhyang201) let me know if I explained it incorrectly. Thanks!
> 
> we would not expect that the best kernel selection changes as seqlen_kv increases (at least the gaps should not be such large). As [@Observer007](https://github.com/Observer007) measured, it makes sense to select attention kernels only based on seqlen_q then.

Agree, then does it make sense to select `trtllm-gen` when seqlen_q == 1 and `cutedsl` when seqlen_q > 1 ? @yhyang201 
> cute-dsl faster above a context threshold 

In my measurement, small context length costs a lot by the host overhead and if we could leverage cuda graph, then cutedsl should be always faster today.

**Comment by yhyang201** [comment]
> Hi, [@yhyang201](https://github.com/yhyang201) Is the performance data number you listed measured by cuda graph? It is a bit weird that in fp8 500k performs better than 131k, below data is measured on B200.
> 
> KV dtype	Q length	Context length	TRTLLM-Gen Eager	TRTLLM-Gen CUDA Graph	CuTe DSL Eager	CuTe DSL CUDA Graph
> BF16	1	131,072	62.4	50.8	101.7	55.3
> BF16	1	500,000	135.8	139.7	185.3	185.9
> BF16	1	1,000,000	264.3	265.2	364.2	364.4
> BF16	8	131,072	206.4	205.1	100.7	62.4
> BF16	8	500,000	725.6	728.2	201.5	204.4
> BF16	8	1,000,000	1,439.7	1,440.4	394.6	393.3
> FP8	1	131,072	63.0	31.8	101.8	32.8
> FP8	1	500,000	70.6	72.9	102.0	95.9
> FP8	1	1,000,000	129.2	131.6	197.7	200.4
> FP8	8	131,072	105.6	108.6	101.4	36.6
> FP8	8	500,000	356.3	360.9	105.5	108.3
> FP8	8	1,000,000	745.5	752.4	204.1	206.0
> cc [@PerkzZheng](https://github.com/PerkzZheng) [@nvpohanh](https://github.com/nvpohanh)

Hi @Observer007, thanks for running this. I can reproduce your numbers on b300, both eager and cuda graph modes. Looks the same on my side: with q=8 and cuda graph capture, cute-dsl is faster than trtllm-gen at every context length, the crossover only exists in eager mode.

Feel free to close this issue. Thank you all. @PerkzZheng @nvpohanh
